### PR TITLE
add test if no loss is provided

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR add a MisConfiguration error when a dictionary is provided by doesn't contain any loss.
If no loss is to provide, don't return anything.

Fixes # (issue)

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -20,20 +20,4 @@ repos:
     rev: v2.3.0
     hooks:
       - id: trailing-whitespace
-      - id: check-yaml
       - id: end-of-file-fixer
-
-  - repo: local
```
Comment (Borda) [inline .pre-commit-config.yaml:26]: why this removal? if so make it as separate PR...

**Review by Borda — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 1 above)
Comment (tchaton) [reply .pre-commit-config.yaml]: We removed them as we decided not to use automatic formatting. So, we just keep the minimum. Will already merged it on master.

Comment (Borda) [reply .pre-commit-config.yaml]: I saw some camouflage commit with the removal but still why to remove it from the pre-commit, if you do not want to happy it, just do not use it... and still why is it part of this PR again?

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -396,6 +394,12 @@ def _process_training_step_output_1_0(self, training_step_output, split_batch):
             hiddens = training_step_output.pop('hiddens', None)
             result['extra'] = training_step_output
 
+            if loss is None:
+                model_ref = self.trainer.get_model()
+                func_name = "training_step_end" if is_overridden('training_step_end', model_ref) else "training_step"
+                m = f'The key `loss` should be present within {func_name} output. Existing keys: {[*training_step_output]}'
+                raise MisconfigurationException(m)
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py:401]:
```suggestion
                raise MisconfigurationException(
                    f'The key `loss` should be present within {func_name} output. Existing keys: {[*training_step_output]}'
                )
```
is there an advantage of having a message as separate var?

**Review by Borda — COMMENTED** [review]
seems like you applied black to the code, can you them mark what are the true changes... 8-)

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -396,6 +394,12 @@ def _process_training_step_output_1_0(self, training_step_output, split_batch):
             hiddens = training_step_output.pop('hiddens', None)
             result['extra'] = training_step_output
 
+            if loss is None:
```
Comment (justusschock) [inline pytorch_lightning/trainer/training_loop.py:397]: don't we support returning None anymore? I thought this was to skip the backward and optimiser step?

Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:397]: Hey @justusschock 

On my understanding of the code, here is what currently supported:

+ None (It will skip backward, optimizer step, etc...)

+ torch.Tensor (Considered as loss)

+ dict 

`current behaviour`:
Pop loss key from dict, create result object with result.minize = loss, and break when creating closure

`new behaviour`:
Pop loss key from dict, make sure loss key with MisConfigurationExpection instead of breaking on closure creation.

 It is a little bit patching legacy code...

Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:397]: @williamFalcon, do you validate with that ?

**Comment by rohitgr7** [comment]
Not sure if it's a good or valid use-case or not. Let's say I want to log images generated by the generator(GAN) in `training_epoch_end` but don't want a backprop at some specific `training_step` due to some reason. So it won't allow me now right?

**Comment by tchaton** [comment]
> Not sure if it's a good or valid use-case or not. Let's say I want to log images generated by the generator(GAN) in `training_epoch_end` but don't want a backprop at some specific `training_step` due to some reason. So it won't allow me now right?

Hey @rohitgr7, you can always return None on training_step. I just check that if you do return a dict, it will need `loss`. It was breaking before, now it is just adding a raise a MisConfiguration.

**Comment by rohitgr7** [comment]
@tchaton if I return None then generated images at that training_step won't be available in training_epoch_end even though I don't want to backprop at the step.

File: tests/trainer/logging/test_train_loop_logging_1_0.py [hunk 4]
```diff
@@ -489,3 +477,70 @@ def validation_step(self, batch, batch_idx):
         weights_summary=None,
     )
     trainer.fit(model, train_data, val_data)
+
+
+def test_misconfiguration_error_for_training_step(tmpdir):
```
Comment (SkafteNicki) [inline tests/trainer/logging/test_train_loop_logging_1_0.py:482]: Could we just parametrize these two test (`@pytest.mark.parametrize('test_train_step_end', [False, True]`)?
Initialize model with
``` python
    class TestModel(BoringModel):
        def training_step(self, batch, batch_idx):
            return {"train_loss": 0}
        def training_step_end(self, out):
            d = {"train_loss": out["train_loss"]}
            return d
```
and then override depending on test arg:
```
model.training_step_end = model.training_step_end if test_train_step_end else None
```

**Comment by williamFalcon** [comment]
@tchaton  yeah, so in manual optimization there should be no requirement to return anything.  and when you do return something it should be allowed (even without a graph attached).

Just because i'm using manual opt, doesn't mean i don't want to have data flowing through the system. So, i guess this exception is likely not something we want to do?

**Comment by tchaton** [comment]
> @tchaton yeah, so in manual optimization there should be no requirement to return anything. and when you do return something it should be allowed (even without a graph attached).
> 
> Just because i'm using manual opt, doesn't mean i don't want to have data flowing through the system. So, i guess this exception is likely not something we want to do?

Hey @williamFalcon, I need to add more tests and definitely check the logic again.

**Comment by tchaton** [comment]
I am closing this PR, as it seems it doesn't aligned with the excepted behaviour. Thanks for your understanding :)
